### PR TITLE
New engine!

### PR DESCRIPTION
**Pull request by chrisnojima**

### CODE REVIEW

File: desktop/app/index.js [hunk 1]
```diff
@@ -1,7 +1,6 @@
 import {BrowserWindow, app, dialog} from 'electron'
 import splash from './splash'
 import installer from './installer'
-import ListenLogUi from '../shared/native/listen-log-ui'
```
Comment (chrisnojima) [inline desktop/app/index.js]: this is now handled in the engine itself

File: protocol/Makefile [hunk 2]
```diff
@@ -33,18 +33,13 @@ js/flow-types.js: build-stamp | config
 	node --harmony ./bin/flow.js
 	cp ./js/flow-types* ../shared/constants/types
 
-js/keybase_v1.js: build-stamp | config
```
Comment (chrisnojima) [inline protocol/Makefile]: the enums have moved to flow-types. Also reduced the hierarchy a level so instead of

```
common.StatusCodes.scgeneric
```

you do

```
CommonStatusCodes.scgeneric
```

File: protocol/bin/flow.js [hunk 3]
```diff
@@ -46,31 +49,74 @@ function load (file, project) {
...
 
 function analyze (json, project) {
+  return reduceArray([].concat(analyzeTypes(json, project), analyzeMessages(json, project), analyzeEnums(json, project)))
+}
+
+function fixCase (s) {
+  return s.toLowerCase().replace(/(_\w)/g, s => capitalize(s[1]))
+}
+
+function analyzeEnums (json, project) {
```
Comment (chrisnojima) [inline protocol/bin/flow.js]: handle enums now

File: protocol/bin/flow.js [hunk 4]
```diff
@@ -92,6 +138,9 @@ function capitalize (s) {
 }
 
 function analyzeMessages (json, project) {
+  // ui means incoming
+  const isUIProtocol = ['notifyCtl'].indexOf(json.protocol) === -1 && !!json.protocol.match(/^notify.*|.*ui$/i)
```
Comment (chrisnojima) [inline protocol/bin/flow.js]: we keep track of whats a ui call or not. this means we can filter out the outgoing calls (the ...rpc()) and the incomingCallMap into the 2 types.

File: protocol/bin/flow.js [hunk 5]
```diff
@@ -254,10 +312,31 @@ export type RPCError = {
...
+type CommonResponseHandler = {
+  error: RPCErrorHandler,
+  result: (...rest: Array<void>) => void,
+}`
 
   const incomingMap = `export type incomingCallMapType = $Exact<{\n` +
-  Object.keys(project.incomingMaps).map(im => `  '${im}'?: ${project.incomingMaps[im]}`).join(',\n') + '\n}>\n\n'
-  s.write(typePrelude + typeDefs.join('') + incomingMap)
+  Object.keys(project.incomingMaps).map(im => `  '${im}'?: ${project.incomingMaps[im]}`).join(',\n') + '\n}>\n'
+  const toWrite = [typePrelude, typeDefs.join('\n\n'), incomingMap].join('\n')
```
Comment (chrisnojima) [inline protocol/bin/flow.js]: do our newlines here to control the spacing easier

File: protocol/bin/flow.js [hunk 6]
```diff
@@ -254,10 +312,31 @@ export type RPCError = {
   code: number,
   desc: string
 }
-`
+export type WaitingHandlerType = (waiting: boolean, method: string, sessionID: number) => void
```
Comment (chrisnojima) [inline protocol/bin/flow.js]: use more helper types here to reduce the amount of code we produce

File: shared/.flowconfig [hunk 7]
```diff
@@ -33,6 +33,7 @@
 munge_underscores=true
 esproposal.class_static_fields=enable
 esproposal.class_instance_fields=enable
+unsafe.enable_getters_and_setters=true
```
Comment (chrisnojima) [inline shared/.flowconfig]: engine uses some getters/setters

File: shared/actions/gregor.js [hunk 8]
```diff
@@ -48,11 +48,9 @@ function registerGregorListeners () {
     })
 
     // we get this with sessionID == 0 if we call openDialog
-    engine.listenGeneralIncomingRpc({
-      'keybase.1.gregorUI.pushState': ({state, reason}, response) => {
-        dispatch(pushState(state, reason))
-        response && response.result && response.result()
-      },
+    engine.setIncomingHandler('keybase.1.gregorUI.pushState', ({state, reason}, response) => {
```
Comment (chrisnojima) [inline shared/actions/gregor.js]: This method takes in a single handler and not an object of them

File: shared/actions/login/index.js [hunk 9]
```diff
@@ -99,7 +100,7 @@ export function login (): AsyncAction {
           param: {
             deviceType,
             usernameOrEmail: usernameOrEmail,
-            clientType: Common.ClientType.gui,
+            clientType: CommonClientType.gui,
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: new enums format

File: shared/actions/login/index.js [hunk 10]
```diff
@@ -110,7 +111,7 @@ export function login (): AsyncAction {
                 payload: error,
               })
 
-              if (!(error.raw && error.raw.code === InputCancelError.code)) {
+              if (!(error.code === InputCancelError.code)) {
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: removed the .raw stuff

File: shared/actions/tracker.js [hunk 11]
```diff
@@ -411,11 +416,11 @@ function updatePGPKey (username: string, pgpFingerprint: Buffer): Action {
   }
 }
 
+const sessionIDToUsername: { [key: number]: string } = {}
 // TODO: if we get multiple tracker calls we should cancel one of the sessionIDs, now they'll clash
-function serverCallMap (dispatch: Dispatch, getState: Function, skipPopups: boolean = false): CallMap {
-  const sessionIDToUsername: { [key: number]: string } = {}
-  const identifyUi = {
-    start: ({username, sessionID, reason}) => {
+function serverCallMap (dispatch: Dispatch, getState: Function, skipPopups: boolean = false, onFinish: ?() => void): incomingCallMapType {
```
Comment (chrisnojima) [inline shared/actions/tracker.js]: removed the call map middleware. this makes this look like everything else and also flowtypes better

File: shared/actions/unlock-folders.js [hunk 12]
```diff
@@ -46,21 +41,13 @@ export function finish (): Finish {
 
 export function openDialog (): AsyncAction {
   return dispatch => {
-    const params: rekeyShowPendingRekeyStatusRpc = {
-      method: 'rekey.showPendingRekeyStatus',
-      callback: null,
-    }
-    engine.rpc(params)
```
Comment (chrisnojima) [inline shared/actions/unlock-folders.js]: dunno why this old style call exists

File: shared/actions/unlock-folders.js [hunk 13]
```diff
@@ -46,21 +41,13 @@ export function finish (): Finish {
...
 }
 
 export function close (): AsyncAction {
   return (dispatch, getState) => {
-    // waiting for cleanupRPC to get merged in so i can make param / sessionID optional
-    // $FlowIssue
-    rekeyRekeyStatusFinishRpc({
-      param: {sessionID: rekeyServer && rekeyServer.sessionID},
-    })
+    rekeyRekeyStatusFinishRpc({})
```
Comment (chrisnojima) [inline shared/actions/unlock-folders.js]: with the rekey refactor on core we don't need to bookkeep the sessionID anymore

File: shared/actions/unlock-folders.js [hunk 14]
```diff
@@ -80,19 +67,17 @@ export function registerRekeyListener (): (dispatch: Dispatch) => void {
...
+    engine.setIncomingHandler('keybase.1.rekeyUI.refresh', (params, response) => refreshHandler(params, response, dispatch))
+
     // else we get this also as part of delegateRekeyUI
-    rekeyServer = createServer(
-      engine,
-      'keybase.1.rekeyUI.delegateRekeyUI',
-      null,
-      () => ({
-        'keybase.1.rekeyUI.delegateRekeyUI': (params, response) => { },
+    engine.setIncomingHandler('keybase.1.rekeyUI.delegateRekeyUI', (param: any, response: ?Object) => {
```
Comment (chrisnojima) [inline shared/actions/unlock-folders.js]: instead of the whole server layer on top of engine we simply listen for this message and create a new session

File: shared/engine/errors.js [hunk 15]
```diff
@@ -1,41 +0,0 @@
-// @flow
```
Comment (chrisnojima) [inline shared/engine/errors.js]: not wrapping this stuff anymore. instead actions/reducers should deal with error code transformations

File: shared/engine/index.js [hunk 16]
```diff
@@ -1,430 +1,284 @@
 // @flow
```
Comment (chrisnojima) [inline shared/engine/index.js]: This has changed a ton and should be looked at as a non-diff

Comment (chrisnojima) [inline shared/engine/index.js]: tldr; Before the engine had maps to keep track of session ids, responses, outstanding requests, etc. Now it just knows about sessions and generalized handlers mostly

File: shared/engine/request.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
+// @flow
```
Comment (chrisnojima) [inline shared/engine/request.js]: Sessions have requests happen

File: shared/engine/transport-shared.js [hunk 18]
```diff
@@ -1,14 +1,85 @@
 // @flow
-import EngineError from './errors'
+// Classes used to handle RPCs. Ability to inject delays into calls to/from server
```
Comment (chrisnojima) [inline shared/engine/transport-shared.js]: delay, logging, only once behavior is now all at this level

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump please

File: shared/engine/index.js [hunk 19]
```diff
@@ -1,435 +1,284 @@
...
-    Object.keys(params).forEach(method => {
-      this.generalListeners[method] = params[method]
+  // Default handlers for incoming messages
+  _setupCoreHandlers () {
+    this.setIncomingHandler('keybase.1.logUi.log', (param, response) => {
+      const logParam: logUiLogRpcParam = param
+      log(logParam)
+      response && response.result && response.result()
     })
+    this.setIncomingHandler('keybase.1.NotifyUsers.userChanged', () => {})
```
Comment (MarcoPolo) [inline shared/engine/index.js]: We should be doing something with this in the future, right?

Comment (chrisnojima) [inline shared/engine/index.js]: unclear

File: shared/engine/index.js [hunk 20]
```diff
@@ -1,435 +1,284 @@
...
+    const {seqid, cancelled} = response || {seqid: 0, cancelled: false}
+    const {sessionID} = param
 
-      if (printOutstandingRPCs) {
-        this.responseMeta[sessionID] = {method, param}
+    if (cancelled) {
+      this._handleCancel(seqid)
+    } else {
+      const session = this._sessionsMap[String(sessionID)]
+      if (session && session.incomingCall(method, param, response)) { // Part of a session?
```
Comment (MarcoPolo) [inline shared/engine/index.js]: Just a note for myself and other reviewers:

This lets the session say "I can't handle this" even if there is a sesssion that matches that session id.

Useful for stuff like pinentry where we don't want each session handling it. It's okay to just fallback

File: protocol/bin/flow.js [hunk 21]
```diff
@@ -26,18 +26,21 @@ var projects = [
   },
 ]
 
+const reduceArray = arr => arr.reduce((acc, cur) => acc.concat(cur), [])
```
Comment (chromakode) [inline protocol/bin/flow.js]: https://lodash.com/docs#flatten ?

**Comment by MarcoPolo** [comment]
not for this PR, but in the future we might want rpc functions able to return something that can be used to cancel them later.

File: protocol/js/flow-types-chat.js [hunk 22]
```diff
@@ -16,6 +16,60 @@ export type RPCError = {
   code: number,
   desc: string
 }
+export type WaitingHandlerType = (waiting: boolean, method: string, sessionID: number) => void
+
+// $FlowIssue we're calling an internal method on engine that's there just for us
+const engineRpcOutgoing = (...args) => engine._rpcOutgoing(...args)
```
Comment (MarcoPolo) [inline protocol/js/flow-types-chat.js]: This is weird that this imports engine for this. Maybe it would be better to have engine be the first argument when you make rpc calls?

It's double weird when you think that importing engine has the side effect of creating a new engine instance. We should probably export a singleton constructor instead.

In the singleton constructor, it doesn't make sense to build the engine here, because that would mean that anytime you want to use types you are building an engine.

I know @gabriel has a couple thoughts about this too.

File: protocol/bin/flow.js [hunk 23]
```diff
@@ -46,31 +49,74 @@ function load (file, project) {
...
+      en[name] = val
+    })
+
+    return {
+      name: `${capitalize(json.protocol)}${t.name}`,
+      map: en,
+    }
+  }).reduce((acc, t) => {
+    return acc.concat([
+      `export const ${t.name} = {
```
Comment (chromakode) [inline protocol/bin/flow.js]: :sheep: `dedent` is pretty nice for this https://www.npmjs.com/package/dedent-js

File: protocol/js/flow-types-chat.js (hunk 22 above)
Comment (chrisnojima) [inline protocol/js/flow-types-chat.js]: so this code already existed. (not that we can't change it). this is so people who are making rpc calls don't need to understand that engine exists else they'll all have to import it just to pass it into this helper (which seems lame).
new engines are not being made, its a singleton thats exported by the module. if we want to export a closure that also exports a singleton thats fine but that isn't a fix for what you're talking about

Comment (gabriel) [inline protocol/js/flow-types-chat.js]: Yeah, I listed some thoughts on that here:
https://github.com/keybase/client/pull/4060

I think constructing (non-simple things) via import and using import to enforce singleton-ness is worth avoiding.

Also especially in this case because engine construction I think triggers an rpc connection.

Not a huge change to make this happen, we can do it after in a separate PR.

Comment (chrisnojima) [inline protocol/js/flow-types-chat.js]: i buy controlling when its constructed. we can force you to do engine() like you suggest. i'm just saying we're not building multiple engines

File: shared/engine/session.js [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+      if (!waiting) {
+        const requests = isOutgoing ? this._outgoingRequests : this._incomingRequests
+        const idx = requests.findIndex(r => r.method === method)
+        if (idx !== -1) {
+          // Mark us as responded
+          if (seqid) {
+            this._seqIDResponded[String(seqid)] = true
+          }
+          // Remove from our list
+          requests.splice(idx, 1)
```
Comment (MarcoPolo) [inline shared/engine/session.js]: not a fan of mutation...

what about 

```
if (isOutgoing) {
 this._outgoingReqs.filter(r => r.method !== method)
} else {
//same
}

if (seqid) {
  this._seqIDResponded[String(seqid)] = true
}
```

File: shared/engine/session.js [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  end () {
+    this._endHandler && this._endHandler(this)
+  }
+
+  // Start the session normally. Tells engine we're done at the end
+  start (method: MethodKey, param: ?Object, callback: ?() => void) {
+    this._startMethod = method
+    this._startCallback = callback
```
Comment (MarcoPolo) [inline shared/engine/session.js]: 🐑 🐑 `startCallback` is a bit confusing of a name, but I can't think of a better one

File: protocol/bin/flow.js (hunk 4 above)
Comment (chromakode) [inline protocol/bin/flow.js]: Perhaps this regex shouldn't be case-insensitive?

File: protocol/bin/flow.js [hunk 26]
```diff
@@ -217,28 +269,31 @@ function parseRecord (t) {
...
-      }, [])
-      .sort()
-      .join('\n  | ')
-
-  console.log(rpcTypes)
+  const rpcTypes = typeDefs.map(t => {
+    const m = t.match(/(\w*Rpc) \(/)
+    return m && m[1]
+  })
+  .filter(t => t)
```
Comment (chromakode) [inline protocol/bin/flow.js]: Filter no longer required here?

File: protocol/bin/flow.js (hunk 4 above)
Comment (chrisnojima) [inline protocol/bin/flow.js]: it has to be to be correct

File: protocol/bin/flow.js (hunk 23 above)
Comment (chrisnojima) [inline protocol/bin/flow.js]: protocol has its own deps and isn't something we touch that often so i'm trying to not have too many dependencies

File: shared/engine/session.js (hunk 24 above)
Comment (chrisnojima) [inline shared/engine/session.js]: i think we want this stuff cleaned up, its our own internal bookkeeping so why have it hang around?

Comment (MarcoPolo) [inline shared/engine/session.js]: actually I see you probably want seqIDResponded only true if you had this in your known requests.   So tweak that demo a bit

File: protocol/js/flow-types-chat.js (hunk 22 above)
Comment (chrisnojima) [inline protocol/js/flow-types-chat.js]: also node framed rpc doens't support sending cancels from our side currently so let's not try and make that work short term

File: shared/engine/transport-shared.js [hunk 27]
```diff
@@ -24,7 +98,38 @@ class TransportShared extends RobustTransport {
...
+
+      // Can't use {...} here due to react-native not doing object.assign on objects w/ proto chains
+      payload.response = {}
+      Object.keys(oldResponse).forEach(key => {
+        payload.response[key] = oldResponse[key]
+      })
+
+      calls.forEach(call => {
+        payload.response[call] = _wrap((...args) => {
+          oldResponse[call](...args)
```
Comment (MarcoPolo) [inline shared/engine/transport-shared.js]: so you don't need the context behind `response.result` anymore?

we can do destructuring in callbacks now?

i.e. `({response: {result}) => result()`

Comment (chrisnojima) [inline shared/engine/transport-shared.js]: i think this is a dangerous pattern and should be avoided. the caller changes binding and now you're broken again

File: shared/engine/session.js (hunk 24 above)
Comment (MarcoPolo) [inline shared/engine/session.js]: This is a 🐑

File: protocol/js/flow-types-chat.js (hunk 22 above)
Comment (MarcoPolo) [inline protocol/js/flow-types-chat.js]: I didn't mean we were building multiple engines. I meant that importing a type has a side effect of building an engine (assuming it didn't exist already).

File: shared/engine/transport-shared.js (hunk 27 above)
Comment (MarcoPolo) [inline shared/engine/transport-shared.js]: you mean `this` is a dangerous pattern ;)

**Comment by MarcoPolo** [comment]
👍 Do I get a 🍪 ?

File: protocol/bin/flow.js (hunk 4 above)
Comment (chrisnojima) [inline protocol/bin/flow.js]: i'll add a comment

**Comment by MarcoPolo** [comment]
👍

File: protocol/bin/flow.js (hunk 26 above)
Comment (chrisnojima) [inline protocol/bin/flow.js]: still needed actually

File: shared/engine/transport-shared.js (hunk 27 above)
Comment (chrisnojima) [inline shared/engine/transport-shared.js]: yeah sadly. the payloads are plain-ol-javascript-objects, but response isn't so destructuring it is gunna be dangerous.

File: shared/actions/pinentry.js [hunk 28]
```diff
@@ -28,8 +27,28 @@ export function registerPinentryListener (): AsyncAction {
       payload: {started: true},
     }: RegisterPinentryListenerAction))
 
-    const pinentryListeners = pinentryListenersCreator(dispatch)
-    engine.listenGeneralIncomingRpc(pinentryListeners)
+    engine.setIncomingHandler('keybase.1.secretUi.getPassphrase', (payload, response) => {
```
Comment (chromakode) [inline shared/actions/pinentry.js]: :sheep: Might be valuable to give this function a name (rather than arrow function) for stack traces

Comment (chrisnojima) [inline shared/actions/pinentry.js]: we don't really do this elsewhere. also with the current stack you can see its a getPassphrase call and thats its a handler so i think its all easy to understand

File: shared/engine/index.js [hunk 29]
```diff
@@ -281,4 +279,12 @@ export type ResponseType = {
   error: (...args: Array<any>) => void,
 }
 
-export default engine
+let engine
+const makeEngine = () => {
```
Comment (chromakode) [inline shared/engine/index.js]: I think this is an improvement on what was before, but one drawback is that the timing of creating the engine is implicit based on first use. This approach introduces a global but defers the organizational task of deciding where to fit it into the namespace and when to initialize it. This is technical debt.

Do we have some other place where shared globals are setup / stored? Should we create a pattern for this?

Comment (chromakode) [inline shared/engine/index.js]: Alternatively a more explicit `setupEngine` called in app startup would be a welcome addition.

File: protocol/bin/flow.js (hunk 26 above)
Comment (chromakode) [inline protocol/bin/flow.js]: Ah I see -- duh!

File: shared/engine/index.js (hunk 19 above)
Comment (chromakode) [inline shared/engine/index.js]: We should add a comment to this line explaining why it's no-opped

File: shared/engine/index.js [hunk 30]
```diff
@@ -1,435 +1,290 @@
...
 
-  _generalIncomingRpc (method: MethodKey, param: Object, response: any) {
-    const listener = this.generalListeners[method]
+  // An outgoing call. ONLY called by the flow-type rpc helpers
+  _rpcOutgoing (params: {
+    method: MethodKey,
+    param?: ?Object,
+    incomingCallMap?: incomingCallMapType,
+    callback?: ?(...args: Array<any>) => void,
+    waitingHandler?: WaitingHandlerType}) {
```
Comment (chromakode) [inline shared/engine/index.js]: It's pretty hard to tell where the arguments end and the functions begin with this style. Maybe drop the closing `) {` to the next line and dedent?

**Comment by chromakode** [comment]
Epic! :+1: % style comments

File: shared/engine/index.js (hunk 19 above)
Comment (chrisnojima) [inline shared/engine/index.js]: done, moved to its own _setupIgnoredHanlders  handler to make things more clear

File: shared/engine/index.js (hunk 29 above)
Comment (chrisnojima) [inline shared/engine/index.js]: i added this. we just init it when we're setting up a new context

**Comment by chrisnojima** [comment]
@MarcoPolo @chromakode https://github.com/keybase/client/pull/3953/commits/c71e3070f3da99db9f3c716f7e34c786ab966a03 addresses your pr comments and then i rebased after. can i get one more thumb then i'll merge

File: shared/engine/index.js [hunk 31]
```diff
@@ -287,10 +289,30 @@ export type ResponseType = {
...
 
-export default makeEngine
+const getEngine = () => {
+  if (__DEV__ && !engine) {
+    throw new Error('Engine needs to be initialized first')
+  }
+
+  // This is just a sanity check so we don't break old code. Should never happen in practice
+  if (!engine) {
+    makeEngine()
```
Comment (chromakode) [inline shared/engine/index.js]: Should we make this throw in dev?

Comment (chromakode) [inline shared/engine/index.js]: Ah I see, handled by `makeEngine` above. NVM!

**Comment by chromakode** [comment]
Updates look good! :+1:

Re: `makeEngine` calls, I didn't notice any for the mobile code. Is that TBD?

**Comment by chrisnojima** [comment]
no lemme fix that tnx

**Comment by chrisnojima** [comment]
sadly i have to update react/react-native on the rn side so we can get pure components for the profile updates

**Comment by chrisnojima** [comment]
@chromakode @MarcoPolo updated w/ mobile fix

**Comment by chromakode** [comment]
:+1:
